toe/src/urbit/app/toe.hoon

```diff
@@ -0,0 +1,921 @@
+        ?:  =(~ out:bc)  [%toe-turno !>(turno:bc)]
+        [%toe-winner !>([(end:co:view [out:bc who]) turno:bc])]
+    ==
+  ::
+  ++  confirm
+    |=  command=tape
+    ^-  (quip card _state)
+    =/  answer=(unit @t)  (rust (cass command) (mask "yn"))
+    ?~  answer  [~ state]
+    =,  room-core
+    ?:(=(u.answer 'n') (close %.n) confirm)
+  --
+::  +room-core: handles actions to a game room
+::
+++  room-core
+  |%
+  ::  %create: appends new room with a player (e.g. ~zod)
+  ::
+  ++  create
+    |=  [guest=@p =game-state]
+    ^-  _state
+    ~&  create+guest
+    state(rooms (snoc rooms [guest ~]), game game-state)
+  ::  %invite: sends an invite to play with opponent (e.g. ~zod)
+  ::
+  ++  invite
+    |=  guest=@p
+    ^-  (list card)
+    =,  view
+    ^-  (list card)
+    :~  [%pass /request %agent [guest %toe] %poke %urbit !>(me)]
+        (waiting:updates:co guest)
+        (waiting:updates:fe guest)
+    ==
+  ::  %receive: a guest ship sends an invite to play/replay
+  ::
+  ++  receive
+    |=  guest=@p
+    ^-  (quip card _state)
+    ~&  active+active
+    =/  enqueue
+      =,  updates:view
+      :-  (receive guest)
+      (create [guest ?~(active %confirm game)])
+    ?~  active  enqueue
+    ?.  =(guest u.active)  enqueue
+    :_  state(next %.y)
+    (notify:updates:view guest)
+  ::  %confirm:  innitializes room state and starts game
+  ::
+  ++  confirm
+    ::  TODO: not only handle oldest room invite
+    ::
+    ^-  (quip card _state)
+    ?>  ?=(^ rooms)
+    =/  ze=@p  ship.i.rooms
+    =/  =toers
+      %-  ~(gas by *toers)
+      ?.(=(%.y ^next) (player-a me ze) (player-b me ze))
+    =/  who=@p
+      ?~  active  me
+      ?:(=(%.y ^next) me ze)
+    =/  request=(list card)
+      ?.  =(game %replay)
+        ~&  in-confirm+[ze /play]
+        [%pass /play %agent [ze %toe] %watch /room/(scot %p me)]~
+      ::  Next is activated if our opponent already confirmed the replay
+      ::
+      ?:  =(%.y next)  ~
+      =/  =toe-player  [%replay [%'X' %g]]
+      [%pass /replay %agent [ze %toe] %poke %toe-player !>(toe-player)]~
+    =.  rooms  [[ze `[*board-game toers who=who]] t.rooms]
+    :_  state(active `ze, game %play)
+    =,  view
+    %+  weld
+      request
+    :~  (playing:updates:co ze)
+        (playing:updates:fe ze)
+    ==
+  ::  +close: leaves the current rooms and closes subscriptions
+  ::
+  ++  close
+    |=  kicked=?
+    ~&  kicked
+    ~&  wex+wex.bol
+    ~&  sup+sup.bol
+    ^-  (quip card _state)
+    ?~  rooms
+       =.  rooms  ~
+       [restart:updates:view wipe]
+    =/  disconnect=(list card)
+      =*  ze  ship:current
+      ?~  room.i.rooms
+        ?:  =(kicked %.y)  ~
+        [%pass /cancel %agent [ze %toe] %poke %toe-cancel !>(%bye)]~
+      %+  weld
+        ::  we leave our subscription
+        ::
+        ~&  in-close+[ze /leave/play]
+        ?.  (~(has by wex.bol) [/play ze %toe])
+          ~&  "none"  ~
```

Comment:
iirc leaving a non-existent subscription is safe, so you don't really need this check. @philipcmonk can probably confirm.

Reply:
Also here and with the casts below, it's generally not good style to put all that tall form onto a single line. Vertical space is "free", use it!

Reply:
The `~(has by wex.bol)` was not there originally, was added after I started seeing the `non existent subscription`, just to check what was in the subscription queue